pkg/cloudprovider/aws.go

```diff
@@ -1,32 +1,342 @@
+// AWS API is separated per family). If the IP is already existing: it returns an
+// AlreadyExistingIPError.
 func (a *AWS) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return err
+	}
+	networkInterfaces, err := a.getNetworkInterfaces(instance)
+	if err != nil {
+		return err
+	}
+	// Perform the operation against the first interface listed following the
+	// order AWS specifies.
+	for _, networkInterface := range networkInterfaces {
+		addIP := ip.String()
+		keepIPs := []*string{}
+		if utilnet.IsIPv6(ip) {
+			for _, assignedIPv6 := range networkInterface.Ipv6Addresses {
+				if assignedIP := net.ParseIP(*assignedIPv6.Ipv6Address); assignedIP != nil && assignedIP.Equal(ip) {
+					return AlreadyExistingIPError
+				}
+				keepIPs = append(keepIPs, assignedIPv6.Ipv6Address)
+			}
+			keepIPs = append(keepIPs, &addIP)
+			input := ec2.AssignIpv6AddressesInput{
+				NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+				Ipv6Addresses:      keepIPs,
+			}
+			_, err = a.client.AssignIpv6Addresses(&input)
+			if err != nil {
+				return err
+			}
+			return a.waitForCompletion(node, awsapi.StringValueSlice(keepIPs), false)
+		}
+		for _, assignedIPv4 := range networkInterface.PrivateIpAddresses {
+			if assignedIP := net.ParseIP(*assignedIPv4.PrivateIpAddress); assignedIP != nil && assignedIP.Equal(ip) {
+				return AlreadyExistingIPError
+			}
+			keepIPs = append(keepIPs, assignedIPv4.PrivateIpAddress)
+		}
+		keepIPs = append(keepIPs, &addIP)
+		inputV4 := ec2.AssignPrivateIpAddressesInput{
+			NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+			PrivateIpAddresses: keepIPs,
+		}
+		_, err = a.client.AssignPrivateIpAddresses(&inputV4)
+		if err != nil {
+			return err
+		}
+		return a.waitForCompletion(node, awsapi.StringValueSlice(keepIPs), false)
+	}
 	return nil
 }
 
+// ReleasePrivateIP un-assigns the IP address from the node by re-assigning the
+// existing IP addresses - the one to be removed.
 func (a *AWS) ReleasePrivateIP(ip net.IP, node *corev1.Node) error {
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return err
+	}
+	networkInterfaces, err := a.getNetworkInterfaces(instance)
+	if err != nil {
+		return err
+	}
+	// Perform the operation against the first interface listed following the
+	// order AWS specifies.
+	for _, networkInterface := range networkInterfaces {
+		deleteIPs := []*string{}
+		if utilnet.IsIPv6(ip) {
+			for _, assignedIPv6 := range networkInterface.Ipv6Addresses {
+				if assignedIP := net.ParseIP(*assignedIPv6.Ipv6Address); assignedIP != nil && assignedIP.Equal(ip) {
+					deleteIPs = append(deleteIPs, assignedIPv6.Ipv6Address)
+				}
+			}
+			if len(deleteIPs) == 0 {
+				return NonExistingIPError
+			}
+			input := ec2.UnassignIpv6AddressesInput{
+				NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+				Ipv6Addresses:      deleteIPs,
+			}
```

Comment:
Do you really need to compute `deleteIPs`? Can't you just do
```
			deleteIP := ip.String()
			input := ec2.UnassignIpv6AddressesInput{
				NetworkInterfaceId: networkInterface.NetworkInterfaceId,
				Ipv6Addresses:      [&deleteIP],
			}
```

and in the unlikely event that it's `NonExistingIPError` you can figure that out from the AWS error?

Reply:
> and in the unlikely event that it's NonExistingIPError you can figure that out from the AWS error?

I don't know, presumably the API should handle it and not return an error if you provide it an empty list of IPs, but I am not sure. I'd prefer we merge the code as is (since I've done some basic testing on all platforms verifying that this works), and I don't think we have time to investigate the cloud API at this point.